File: tools/bench/vii3_parametric_vs_linear.cpp
// VII-3 본문 수치 — 답을 선형으로 훑는 것과 이분 탐색하는 것의 차이 (C++ 판).
// Python 판(vii3_parametric_vs_linear.py)과 같은 입력·같은 판정 함수를 쓴다.
// 입력은 난수가 아니라 결정적 수열이라 두 언어의 답이 정확히 같아야 한다.
//
// 측정 환경은 CLAUDE.md §1-3.
// 빌드/실행:
//   g++ -std=c++17 -O2 tools/bench/vii3_parametric_vs_linear.cpp -o /tmp/vii3 && /tmp/vii3

#include <chrono>
#include <cstdio>
#include <algorithm>
#include <vector>
using namespace std;
using namespace std::chrono;

const int K = 100;
const int HI = 50000;
const int N = 137;
const int REPEAT = 5;

vector<int> a;
long long calls = 0;

// 길이 x 로 자르면 N 개 이상 나오는가. x 가 커질수록 참에서 거짓으로 한 번 뒤집힌다.
bool feasible(int x) {
    calls++;
    long long total = 0;
    for (int v : a) total += v / x;
    return total >= N;
}

int solve_linear() {
    int best = 0;
    for (int x = 1; x < HI; x++) {
        if (feasible(x)) best = x;
        else break;
    }
    return best;
}

int solve_binary() {
    int lo = 1, hi = HI;
    while (lo < hi) {
        int mid = lo + (hi - lo) / 2;
        if (!feasible(mid)) hi = mid;
        else lo = mid + 1;
    }
    return lo - 1;
}

template <class F>
pair<int, double> timed(F fn) {
    vector<double> ts;
    int ans = 0;
    for (int i = 0; i < REPEAT; i++) {
        auto t0 = steady_clock::now();
        ans = fn();
        auto t1 = steady_clock::now();
        ts.push_back(duration<double>(t1 - t0).count());
    }
    sort(ts.begin(), ts.end());
    return {ans, ts[ts.size() / 2]};
}

int main() {
    // 난수 대신 결정적 수열을 쓴다. 두 언어의 난수 생성기가 달라도 입력이 같아야
    // 표의 두 열을 나란히 놓을 수 있다.
    a.resize(K);
    for (int i = 0; i < K; i++) a[i] = (i * 7919 + 13) % (HI - 1) + 1;

    calls = 0;
    auto lin = timed(solve_linear);
    long long calls_lin = calls / REPEAT;

    calls = 0;
    auto bin = timed(solve_binary);
    long long calls_bin = calls / REPEAT;

    printf("K = %d  HI = %d  N = %d\n", K, HI, N);
    printf("선형  답=%d  판정 호출=%7lld  %9.3f ms\n", lin.first, calls_lin, lin.second * 1e3);
    printf("이분  답=%d  판정 호출=%7lld  %9.3f us\n", bin.first, calls_bin, bin.second * 1e6);
    printf("두 답이 같은가: %s\n", lin.first == bin.first ? "true" : "false");
    printf("판정 호출 비 = %.1f배, 시간 비 = %.1f배\n",
           (double)calls_lin / calls_bin, lin.second / bin.second);
    return 0;
}

File: tools/bench/vii3_parametric_vs_linear.py
"""VII-3 본문 수치 — 답을 선형으로 훑는 것과 이분 탐색하는 것의 차이.

같은 판정 함수 feasible(x) 를 쓴다. 다른 것은 x 를 고르는 순서뿐이다.
선형은 답 근처까지 x 를 하나씩 올리며 판정하고, 이분은 구간을 절반씩 접는다.
판정 비용이 O(K) 이므로 총비용은 (판정 횟수) x K 다 — 그 곱을 눈으로 보이는 것이 목적이다.

측정 환경은 CLAUDE.md §1-3.
실행:  python3.13 tools/bench/vii3_parametric_vs_linear.py
"""

import time

K = 100          # 랜선 개수. 판정 한 번의 비용이 곧 K 다
HI = 50_000      # 답 후보 구간의 오른쪽 끝 (반열림)
N = 137          # 만들어야 하는 랜선 개수
REPEAT = 5

# 난수 대신 결정적 수열을 쓴다. 두 언어의 난수 생성기가 달라도 입력이 같아야
# 표의 두 열을 나란히 놓을 수 있다.
a = [(i * 7919 + 13) % (HI - 1) + 1 for i in range(K)]

calls = 0


def feasible(x):
    """길이 x 로 자르면 N 개 이상 나오는가. x 가 커질수록 참에서 거짓으로 한 번 뒤집힌다."""
    global calls
    calls += 1
    total = 0
    for v in a:
        total += v // x
    return total >= N


def solve_linear():
    """x 를 1 부터 올리며 마지막 참을 기억한다. 판정 횟수가 곧 답의 크기다."""
    best = 0
    for x in range(1, HI):
        if feasible(x):
            best = x
        else:
            break          # 단조라서 처음 거짓에서 멈춰도 된다. 그래도 답 크기만큼 돈다
    return best


def solve_binary():
    """p(x) = not feasible(x) 는 F...FT...T 다. 첫 참을 찾고 한 칸 왼쪽이 답."""
    lo, hi = 1, HI
    while lo < hi:
        mid = lo + (hi - lo) // 2
        if not feasible(mid):
            hi = mid
        else:
            lo = mid + 1
    return lo - 1


def timed(fn):
    best = None
    times = []
    for _ in range(REPEAT):
        t0 = time.perf_counter()
        best = fn()
        times.append(time.perf_counter() - t0)
    times.sort()
    return best, times[len(times) // 2]


calls = 0
ans_lin, t_lin = timed(solve_linear)
calls_lin = calls // REPEAT

calls = 0
ans_bin, t_bin = timed(solve_binary)
calls_bin = calls // REPEAT

print(f"K = {K}  HI = {HI}  N = {N}")
print(f"선형  답={ans_lin}  판정 호출={calls_lin:>7}  {t_lin * 1e3:9.3f} ms")
print(f"이분  답={ans_bin}  판정 호출={calls_bin:>7}  {t_bin * 1e6:9.3f} us")
print(f"두 답이 같은가: {ans_lin == ans_bin}")
print(f"판정 호출 비 = {calls_lin / calls_bin:.1f}배, 시간 비 = {t_lin / t_bin:.1f}배")

# 실제 제약(길이 <= 2^31-1)에서 이분 탐색이 부르는 판정 횟수
lo, hi, cnt = 1, 2**31, 0
while lo < hi:
    mid = lo + (hi - lo) // 2
    cnt += 1
    lo = mid + 1 if mid < 1_400_000_000 else lo
    if mid >= 1_400_000_000:
        hi = mid
print(f"구간 폭 2^31 일 때 이분 탐색의 판정 호출 = {cnt}회")
